Resolve inconsistencies in using controlled gates & controlled operations

tanujkhattar opened the pull request:
This PR tries to resolve the inconsistencies mentioned in #4172. Specifically, 

- Equality b/w controlled `TOFFOLI`s with different qubit ordering on controls, i.e. `TOFFOLI(a,b,c).controlled_by(d)  == TOFFOLI(d,b,c).controlled_by(a)`  (and other 3Q controlled gates like CCZ, CSWAP). To achieve this, we override the `controlled` method on `CCX`, `CCZ` etc. to return a `ControlledGate` with `sub_gate = X` in case of `CCX` s.t. `TOFFOLI(a,b,c).controlled_by(d) == CCCX(a, b, d, c) == TOFFOLI(d,b,c).controlled_by(a)` instead of `CTOFFOLI`
- `gate_operation.controlled_by` now forwards the request to `gate.controlled` to first create a controlled gate and then apply it on qubits to create a `ControlledOperation`. This solves the original use case of adding specialized controls, requested in #2142, i.e. `cirq.Z(q0).controlled_by(q1) == cirq.CZ(q1, q0)`. 
- Fixes #4515


Note, this is a breaking change because
- `gate_operation.controlled_by()` can now return a `cirq.GateOperation` instead of `cirq.ControlledOperation` in cases where the underlying gates have specialized `gate.controlled()` implementations. 
  - This also leads to a change in diagram of the controlled gates with specialized controlled implementations. For eg: Controlled S gate is now plotted as `CZPowGate` (`@ --- @ ** 0.5`) instead of `ControlledOperation` with Z ** 0.5 as subgate(`@ ---- S`)
- `op.controlled_by` for 3Q gates like `CCX`, `CCZ`,  `CSWAP` will now return `ControlledOperation` with `sub_operation = <underlying non-controlled gate>`. Eg: `CCCX` (i.e. `sub_gate = X`) instead of `CTOFFOLI` (i.e. `sub_gate = TOFFOLI`) etc.
- Diagrams for `ControlledOperations` will now always have the exponent drawn on the target qubit (in case of multi qubit `sub_operation`, the exponent will always be on the first qubit if not the underlying gate does not explicitly specify a target index).

balopat reviewed (commented):
Two comments for discussion before we move forward: 

1. This is technically a breaking change as user code might depend on the type of the `controlled_by` method being a `ControlledGate`. So we should make it very explicit in our release notes and label this PR as a breaking change. 
2. Is there really no value at all in subsequent layers of `ControlledGate`s? It feels like a very generic structure that maybe some decomposition code could leverage. But with jumping directly to "specialized" implementation we lose this structure. Maybe could have both? I.e. from a (multi-)controlled gate we could have a method `reduce_type`, `simplify_type` or `specialize_type`  that could return the equivalent type in the hierarchy as needed based on the underlying gate if there is one...I.e. we would do the delegation to the subgate only on demand - this pattern would be similar to `simplify` Sympy expressions.

tanujkhattar commented:
Thanks for the feedback. I found some more cases and have opened #4172 to discuss the various issues related to controlled gates & operations.
I'll mark this PR as draft for now till the discussion on the issue is resolved.

tanujkhattar commented:
@balopat  This is ready for review. I'm open to suggestions for a better breaking change description in the PR.

balopat reviewed (changes requested):
I feel like that the consistency tests should follow the pattern in `cirq.testing.assert_implements_consistent_protocols` -- i.e. I'd create an assertion in cirq.testing that takes a gate, and as sequence of pows, and then do the two consistency tests. Then I'd use this assert method for all of our gates to achieve full coverage across all our gates.

balopat commented:
Re: breaking change note - looks good, but I would specifically mention in the breaking changes the consequent change in QFT and quirk diagrams controlled operations. Some people depend on diagrams for their tests...

tanujkhattar commented:
@balopat I've modified the PR description and added `assert_controlled_and_controlled_by_identical` as suggested. PTAL. 

Also, do you want me to add 100% coverage across all Cirq gates for `assert_controlled_and_controlled_by_identical` in this PR itself? Or could it be a follow-up PR ?

tanujkhattar commented:
@balopat  Ping, this is ready for re-review. PTAL.

tanujkhattar commented:
@MichaelBroughton Can you please take a look at this?

tanujkhattar commented:
@MichaelBroughton Bump!

Strilanc commented on cirq-core/cirq/interop/quirk/cells/input_rotation_cells_test.py:
```diff
@@ -89,9 +89,9 @@ def test_input_rotation_cells():
         '{"cols":[["•","X^(-A/2^n)","inputA2"]]}',
         diagram="""
-0: ───@────────────
+0: ───@^(-A/2^2)───
```
Why did the exponent move from the target to the ancilla? That's not where it should be.

tanujkhattar replied:
This is because the gate changed from being a `ControlledOperation(sub_type='X')` to a `CX` gate. The latter always prints the exponent on the large qubit, and not necessarily on the target. Eg:

```python
In [1]: import cirq
   ...: q = cirq.LineQubit.range(2)
   ...: gate = cirq.CX ** 0.25
   ...: c1 = cirq.Circuit(gate(q[0], q[1]))
   ...: c2 = cirq.Circuit(gate(q[1], q[0]))
   ...: print(c1, c2, sep = '\n---------\n')
0: ───@────────
      │
1: ───X^0.25───
---------
0: ───X────────
      │
1: ───@^0.25───
```

tanujkhattar replied:
Sent another PR to fix the behavior of `CX` and `CCX` gates - https://github.com/quantumlib/Cirq/pull/4462

Strilanc replied:
I don't want to break then unbreak the diagrams. Merge the relevant changes from the other PR into this one. Make zero breaking changes instead of two breaking changes that undo each other.

tanujkhattar replied:
Once the other PR is merged and this PR is rebased, this particular breaking change would go away.

Also, the two PRs / changes would independently always be breaking changes. We can't merge them into one PR and not have a breaking change. This is because:
1) The other PR breaks all diagrams using CX and CCX gates. This includes diagrams inside `cirq-core/cirq/ops/common_gates_test.py` which have nothing to do with this PR. 
2) This PR breaks diagrams where a `@ ---- S` would get plotted as `@ --- @ ** 0.5`. These breaking changes are also unavoidable and have nothing to do with the other PR. 

The other PR is approved and I'll merge once this discussion is resolved.

maffoo replied:
I think the diagrammatic representations of gates are important and it's good that we have tests to ensure that we don't break them, but I don't think we should consider the diagrams to be part of the _stable_ interface for which we provide backwards compatibility guarantees. This strikes me as something like making guarantees about error messages so that users can match errors with regexes and we won't break them. That seems like an anti-pattern; users should match on error _types_ instead. And similarly cirq users should IMO check types and values in their tests, not string diagrams (unless they are testing their own diagram code for their own gates). I realize this means us, since we have a lot of internal tests that match circuit diagrams, but I think we should change that or be willing to update tests when upstream diagrams change.

tanujkhattar replied:
@Strilanc Ping. LMK your thoughts so we can merge the other pending PR and unblock this one as well.

tanujkhattar replied:
@Strilanc Ping. Gentle reminder.

tanujkhattar replied:
I submitted #4462  but there are more bugs related to drawing diagrams for controlled operations in other classes. See #4515

Strilanc commented on cirq-core/cirq/interop/quirk/url_to_circuit_test.py:
```diff
@@ -371,21 +371,21 @@ def test_completes_weight_zero_billion_laughs():
 
 def test_example_qft_circuit():
     qft_example_diagram = """
```
We should avoid making these kinds of diagrammatic changes, precisely because people use them in tests that need to be updated when we do so.

tanujkhattar replied:
I agree it's a breaking change, but I think it's unavoidable here since we are changing the type returned by `controlled_by`.

Strilanc commented on cirq-core/cirq/ops/raw_types_test.py:
```diff
@@ -127,13 +127,13 @@ def test_gate():
+    a, b, c, d = cirq.LineQubit.range(4)
     g = ValiGate()
-    op = g(a)
-    assert op.controlled_by() is op
```
Why has this test changed what it asserts? The method no longer returns the same value when adding zero controls?

tanujkhattar replied:
It returns the same value but didn't return `self` because in the new implementation of `controlled_by` on `GateOperation`, I directly forwarded the request to the underlying `gate` instead of having an explicit check of the form.
```python
    if len(control_qubits) == 0:
        return self
```

Added the check and reverted the test from `==` to `is`.

Strilanc commented on cirq-core/cirq/ops/three_qubit_gates_test.py:
```diff
@@ -319,3 +319,24 @@ def test_resolve(resolve_fn):
+    ],
+)
+def test_controlled_gate_op_consistent(gate):
+    cirq.testing.assert_controlled_and_controlled_by_identical(gate)
```
Could this be part of the standard gate test battery, instead of being performed special for these specific gates?

tanujkhattar replied:
Can you elaborate a bit on what you mean by the standard gate test battery?

smitsanghavi replied:
Was going to comment the same thing. This could be a part of `assert_implements_consistent_protocols`.

MichaelBroughton reviewed (commented):
Thanks @Strilanc for taking a look.  Looking at how things would behave now:
```python3
>>> cirq.X.controlled(3)
cirq.ControlledGate(sub_gate=cirq.X, num_controls=3)
>>> cirq.X.controlled(2)
cirq.TOFFOLI
>>> cirq.X.controlled(1)
cirq.CNOT
>>> cirq.CNOT.controlled(3)
cirq.ControlledGate(sub_gate=cirq.X, num_controls=4)
>>> cirq.CNOT.controlled(2)
cirq.ControlledGate(sub_gate=cirq.X, num_controls=3)
>>> cirq.CNOT.controlled(1)
cirq.TOFFOLI
```
Seems a little scary type-wise to me, but I guess is more intuitive in some senses. The only thing I would strongly suggest for this PR is to keep the `controlled_by` and `controlled` promotion / demotion behavior tests local to the gates themselves and not make a separate module in `testing` if possible.

Strilanc commented:
I disagree that the testing should be specific to the gates and not in `testing`. `controlling` is a global sort of mutation, and so deserves a global sort of validation that it is consistent.

tanujkhattar commented:
@Strilanc @MichaelBroughton Ping. This is ready for re-review.

Strilanc reviewed (dismissed):
Don't make any breaking changes to diagrams in this PR.

tanujkhattar commented:
Also added a fix for #4515 in this PR to avoid breaking the diagrams twice. See https://github.com/quantumlib/Cirq/pull/4529#issuecomment-926984204 for justification. 

This PR is again ready for re-review

MichaelBroughton reviewed (approved):
I'm ready to merge on this. @maffoo or @smitsanghavi , any final thoughts ?

smitsanghavi reviewed (approved):
LGTM
